### DISCUSSION TITLE
Ways to handle entitlements with SPM

### DISCUSSION THREAD

**Discussion by Lutzifer** [cirruslabs/tart#85] [General]
As of now, building the project needs an additional script call to include the necessary entitlements.

```
codesign --sign - --entitlements Resources/tart.entitlements --force tart
```

I did not find yet any native way to set the entitlements file using Package.swift

Maybe this is a good feature request for the SPM team?

**Comment by edigaryev** [comment]
Sounds like a good idea, this was actually mentioned in the [SE-0303 proposal's "Future Directions" section](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md#future-directions):

>the ability to define commands that run after building and testing is complete

The proposal also gives a hint on [how this can be achieved](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md#using-command-to-postprocess-built-artifacts) using the regular build commands:

>This proposal focuses on tools that run during the build, as part of constructing artifacts. A future proposal could define new plugin capabilities for more general actions that should take place at the very end of a build, after tests have completed.

>Regular build commands can be used for this purpose if they are configured with artifacts inside the built-products directory as their inputs.

However, emitting a `Command.buildCommand()` with it's `inputFiles` set to the `tart` binary in `.build` results in `error: build cycle detected`, so no luck here.

**Reply by Lutzifer** [reply]
yeah, I also ran into the cycle issue.

**Comment by fkorotkov** [comment]
I initially tried to do it in [`swift-package` branch](https://github.com/cirruslabs/tart/blob/swift-package/Package.swift) via a plugin. In theory it should work but unfortunately [this if statement](https://github.com/cirruslabs/tart/blob/24c366a25897aec3498784b3439aa1045c580053/Plugins/SwiftEntitlementsPlugin/plugin.swift#L5) is never true even though `BinaryArtifactTarget`'s documentation states:

> Represents a target describing an artifact (e.g. a library or executable) that is distributed as a binary.

Which I assumed should be true for `tart` executable target but it's not. Might be a good idea to create an issue in SPM repo. When I initially tried it the plugin system was just released.

**Comment by Lutzifer** [comment]
Thanks for your answers, I learned a lot about SwiftPM today 😆

Sounds like waiting for something like https://github.com/apple/swift-evolution/blob/main/proposals/0332-swiftpm-command-plugins.md is necessary.
